File: go/internal/harness/hist.go
package harness

import "math/bits"

// Histogram is a log-linear latency histogram ("loglin6"): values below 64 get
// exact buckets; above that every power of two is split into 64 linear
// sub-buckets, bounding the relative bucket width at 1/64 (~1.6%). The Rust
// harness implements the identical scheme; scripts/benchctl decodes it.
type Histogram struct {
	counts        [histBuckets]uint64
	count         uint64
	sum, min, max uint64
}

const (
	histSubBits = 6
	histSub     = 1 << histSubBits
	histBuckets = (64 - histSubBits + 1) * histSub // 3776
)

// HistBucket returns the bucket index of v.
func HistBucket(v uint64) int {
	if v < histSub {
		return int(v)
	}
	e := 63 - bits.LeadingZeros64(v) // >= histSubBits
	m := (v >> uint(e-histSubBits)) & (histSub - 1)
	return (e-histSubBits+1)*histSub + int(m)
}

// Record adds one observation (nanoseconds).
func (h *Histogram) Record(v uint64) {
	h.counts[HistBucket(v)]++
	if h.count == 0 || v < h.min {
		h.min = v
	}
	if v > h.max {
		h.max = v
	}
	h.count++
	h.sum += v
}

func (h *Histogram) export() map[string]any {
	buckets := make([][2]uint64, 0, 64)
	for i, c := range h.counts {
		if c != 0 {
			buckets = append(buckets, [2]uint64{uint64(i), c})
		}
	}
	return map[string]any{
		"scheme":  "loglin6",
		"count":   h.count,
		"sum_ns":  h.sum,
		"min_ns":  h.min,
		"max_ns":  h.max,
		"buckets": buckets,
	}
}

File: rust/common/src/hist.rs
//! Log-linear latency histogram ("loglin6"), identical to
//! go/internal/harness/hist.go: exact buckets below 64, then 64 linear
//! sub-buckets per power of two (relative bucket width <= 1/64).

use serde_json::{Value, json};

const SUB_BITS: u32 = 6;
const SUB: u64 = 1 << SUB_BITS;
pub const BUCKETS: usize = ((64 - SUB_BITS + 1) as u64 * SUB) as usize; // 3776

/// Bucket index of `v`.
#[inline]
pub fn bucket(v: u64) -> usize {
    if v < SUB {
        return v as usize;
    }
    let e = 63 - v.leading_zeros(); // >= SUB_BITS
    let m = (v >> (e - SUB_BITS)) & (SUB - 1);
    ((e - SUB_BITS + 1) as u64 * SUB + m) as usize
}

pub struct Histogram {
    counts: Vec<u64>,
    count: u64,
    sum: u64,
    min: u64,
    max: u64,
}

impl Default for Histogram {
    fn default() -> Self {
        Self::new()
    }
}

impl Histogram {
    pub fn new() -> Self {
        Histogram { counts: vec![0; BUCKETS], count: 0, sum: 0, min: 0, max: 0 }
    }

    #[inline]
    pub fn record(&mut self, v: u64) {
        self.counts[bucket(v)] += 1;
        if self.count == 0 || v < self.min {
            self.min = v;
        }
        if v > self.max {
            self.max = v;
        }
        self.count += 1;
        self.sum += v;
    }

    pub fn export(&self) -> Value {
        let buckets: Vec<[u64; 2]> =
            self.counts.iter().enumerate().filter(|(_, c)| **c != 0).map(|(i, c)| [i as u64, *c]).collect();
        json!({
            "scheme": "loglin6",
            "count": self.count,
            "sum_ns": self.sum,
            "min_ns": self.min,
            "max_ns": self.max,
            "buckets": buckets,
        })
    }

    pub fn count(&self) -> u64 {
        self.count
    }
}
